sirepo/pkcli/runner.py

```diff
@@ -105,122 +93,192 @@ def _write_status(status, run_dir):
-        # the on-disk simulation_db versus the in-memory status.
-        self.jobs = {}
+        self.report_jobs = {}
+        self.locks = _LockDict()
         self._nursery = nursery
 
-    def status(self, run_dir, jhash):
+    def run_dir_status(self, run_dir):
+        """Get the current status of whatever's happening in run_dir.
+
+        Returns:
+          Tuple of (jhash or None, status of that job)
+
+        """
         disk_in_path = run_dir.join('in.json')
         disk_status_path = run_dir.join('status')
         if disk_in_path.exists() and disk_status_path.exists():
+            # status should be recorded on disk XOR in memory
+            assert run_dir not in self.report_jobs
             disk_in_text = pkio.read_text(disk_in_path)
             disk_jhash = pkjson.load_any(disk_in_text).reportParametersHash
-            if disk_jhash == jhash:
-                disk_status = pkio.read_text(disk_status_path)
-                if disk_status == 'pending':
-                    # We never write this, so it must be stale, in which case
-                    # the job is no longer pending...
-                    pkdlog(
-                        'found "pending" status, treating as "error" ({})',
-                        disk_status_path,
-                    )
-                    return runner_client.JobStatus.ERROR
-                return runner_client.JobStatus(disk_status)
-        if run_dir in self.jobs and self.jobs[run_dir].jhash == jhash:
-            return self.jobs[run_dir].status
-        return runner_client.JobStatus.MISSING
-
-    async def start_job(self, run_dir, jhash, cmd):
-        await self._nursery.start(self._run_job, run_dir, jhash, cmd)
-
-    async def _run_job(
-            self, run_dir, jhash, cmd, *, task_status=trio.TASK_STATUS_IGNORED
-    ):
-        # XX TODO: there are still some awkward race conditions here if a new
-        # job tries to start using the directory while another job is still
-        # using it. probably start_job should detect this, and either kill the
-        # existing job (if it has a different jhash + older serial), do
-        # nothing and report success (if the existing job has the same jhash),
-        # or error out (if the existing job has a different jhash + newer
-        # serial).
-        with _catch_and_log_errors(Exception, 'error in run_job'):
-            if run_dir in self.jobs:
-                # Right now, I don't know what happens if we reach here while
-                # the previous job is still running. The old job might be
-                # writing to the new job's freshly-initialized run_dir? This
-                # will be fixed once we move away from having server.py write
-                # directly into the run_dir.
+            disk_status = pkio.read_text(disk_status_path)
+            if disk_status == 'pending':
+                # We never write this, so it must be stale, in which case
```

Comment:
In `run_dir_status` we don't know if the user wants to run the job again or what. But in `api_runSimulation`, if the current status is `ERROR` then we do re-start the job.

We use the status file to remember the state of already-finished jobs, that the runner isn't otherwise tracking anymore. By the time we save the `disk_status`, the `run_dir` is basically immutable, so there's no point in keeping in-memory data structures. Technically we don't need to do it this way, but we need some way to spill the status to disk to get it out of RAM, and this is good as anything – plus I'm pretty sure there's still code in `simulation_db` that expects the status file to be there.